+nosnoc/+discrete_time_problem/Stewart.m

```diff
@@ -376,29 +410,49 @@ function generate_direct_transcription_constraints(obj)
                     obj.f = obj.f + dcs.f_q_fun(x_ijk, z_ijk, lambda_ijk, theta_ijk, mu_ijk, u_i, v_global, p);
                 end
 
-                % Clock <Constraints
+                % Clock Constraints
                 % TODO(@anton) HERE BE DRAGONS. This is by far the worst part of current nosnoc as it requires the discrete problem
-                %              to understand something about the
-                %              time-freezing reformulation which is ugly. (It is also just a quite complicated thing)
-                if obj.opts.use_fesd && opts.equidistant_control_grid
-                    if opts.time_optimal_problem
-                        if opts.use_speed_of_time_variables
-                            obj.g.equidistant_control_grid(ii) = {[sum_h - opts.h;s_sot*sum_h - obj.w.T_final()/opts.N_stages]};
-                        else
-                            obj.g.equidistant_control_grid(ii) = {sum_h - obj.w.T_final()/opts.N_stages};
-                        end
+                %              to understand something about the time-freezing reformulation which is ugly.
+                % handle numerical time
+                if opts.use_fesd && opts.equidistant_control_grid
+                    relax = vdx.RelaxationStruct(opts.relax_terminal_numerical_time.to_vdx, 's_numerical_time', 'rho_numerical_time');
+                    if opts.time_optimal_problem && ~opts.time_freezing
+                        ecg_rhs = obj.w.T_final()/opts.N_stages;
                     else
-                        if opts.relax_terminal_numerical_time
-                            % Negative values sometimes help convergence.
-                            obj.w.s_numerical_time(ii) = {{'s_numerical', 1}, -2*opts.h, 2*opts.h, opts.h/2};
-                            g_eq_grid = [sum_h - t_stage - obj.w.s_numerical_time(ii);
-                                -(sum_h - t_stage) - obj.w.s_numerical_time(ii)];
-                            obj.g.equidistant_control_grid(ii) = {g_eq_grid, -inf, 0};
-                            obj.f = obj.f + opts.rho_terminal_numerical_time*obj.w.s_numerical_time(ii);
+                        ecg_rhs = obj.p.T()/opts.N_stages;
+                    end
+
+                    if opts.use_numerical_clock_state
+                        curr_t = obj.w.numerical_time(ii,opts.N_finite_elements(ii));
+                        obj.g.equidistant_numerical_grid(ii) = {curr_t - ii*ecg_rhs, relax};
+                    else
+                        if ~opts.time_freezing
+                            obj.g.equidistant_numerical_grid(ii) = {s_sot*sum_h - ecg_rhs, relax};
                         else
-                            obj.g.equidistant_control_grid(ii) = {t_stage-sum_h};
+                            obj.g.equidistant_numerical_grid(ii) = {sum_h - ecg_rhs, relax};
                         end
                     end
+                    if relax.is_relaxed
```

Comment:
explain syntax and consider using more instructive names.
relax.is_relaxed is maybe nonobvious